### PR TITLE
fuzz console log & test cases

### PR DESCRIPTION
**Pull request by Azleal**
## Motivation
As discussed in  #3844 , when using fuzz test, console.log doesn't work as expected(display console log info). This PR added a new fuzz config `show_execution_logs`, with which user could decide if the full console log should be displayed. by defaut, the value is `false`, so it won't affect any fuzz test prior to this change will be made.

## Solution
A `logs` field is added to `CaseOutcome` to keep track of logs that generated in a single fuzz test. And collect logs in `FuzzTestData` after each fuzz test finished. then use the config `show_logs` in `FuzzConfig` to decide if full console log output should be displayed. if not, the `console.log` behavior is the same as what it is now.

Also, you can use `/// forge-config: default.fuzz.show-logs = true` to enable fuzz test console log per test.

4 test cases is added, to test if the new config works as expected.

### CODE REVIEW

**Comment by Azleal** [comment]
hi @grandizzy ,  please review this, there's a failed test case in test integration, and i believe it has nothing to do with my pr. btw, I think the [docs](https://book.getfoundry.sh/reference/config/testing?highlight=fuzz#fuzz) should be updated if the pr is merged. but I didn't find where to update the docs. please tell me if this work needs to be done, and i'll update it.

File: crates/evm/evm/src/executors/fuzz/mod.rs [hunk 1]
```diff
@@ -90,6 +90,10 @@ impl FuzzedExecutor {
         ];
         // We want to collect at least one trace which will be displayed to user.
         let max_traces_to_collect = std::cmp::max(1, self.config.gas_report_samples) as usize;
+        let show_execution_logs = self.config.show_execution_logs;
+
+        //Stores logs for all fuzz cases
+        let logs: RefCell<Vec<Log>> = RefCell::default();
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/fuzz/mod.rs:96]: Let's not store logs but print them right away (as requested in #3844 )

**Review by grandizzy — CHANGES_REQUESTED** [review]
Thank you, nice progress!

I was thinking more along the lines of https://github.com/foundry-rs/foundry/issues/3844 requirement, that is to print logs during the test rather than after the test is completed. That's good for long campaigns where one would like to get feedback. So instead storing logs for all fuzz cases (which would also require more resources for executing these test) we can print logs right after the executor call (if `show-execution-logs` enabled), smth like adding
```Rust
        if self.config.show_execution_logs {
            for log in decode_console_logs(&call.logs) {
                println!("  {log}");
            }
        }
```
right after

https://github.com/foundry-rs/foundry/blob/15652dc9ac56fa12310a901a37489e550ebb133c/crates/evm/evm/src/executors/fuzz/mod.rs#L213-L222

Would be nice to extend this to invariant testing as well (would probably require the config to be a general one like `show_fuzz_logs` and apply to both fuzz and invariant) but can be handled in a separate PR.

**Comment by Azleal** [comment]
> Thank you, nice progress!
> 
> I was thinking more along the lines of #3844 requirement, that is to print logs during the test rather than after the test is completed (good for long campaigns where one would like to get feedback during execution). So instead storing logs for all fuzz cases (which would also require more resources for executing these test) we can print logs right after the executor call (if `show-execution-logs` enabled), smth like adding
> 
> ```rust
>         if self.config.show_execution_logs {
>             for log in decode_console_logs(&call.logs) {
>                 println!("  {log}");
>             }
>         }
> ```
> 
> right after
> 
> https://github.com/foundry-rs/foundry/blob/15652dc9ac56fa12310a901a37489e550ebb133c/crates/evm/evm/src/executors/fuzz/mod.rs#L213-L222
> 
> Would be nice to extend this to invariant testing as well (would probably require the config to be a general one like `show_fuzz_logs` and apply to both fuzz and invariant) but can be handled in a separate PR.

ok, I got your points. Although `show-execution-logs` is under `fuzz` config section, it's clearer to use `show_fuzz_logs` and this convention could also be applied to `invariant` tests. I'm on this.

**Comment by Azleal** [comment]
> Thank you, nice progress!
> 
> I was thinking more along the lines of #3844 requirement, that is to print logs during the test rather than after the test is completed (good for long campaigns where one would like to get feedback during execution). So instead storing logs for all fuzz cases (which would also require more resources for executing these test) we can print logs right after the executor call (if `show-execution-logs` enabled), smth like adding
> 
> ```rust
>         if self.config.show_execution_logs {
>             for log in decode_console_logs(&call.logs) {
>                 println!("  {log}");
>             }
>         }
> ```
> 
> right after
> 
> https://github.com/foundry-rs/foundry/blob/15652dc9ac56fa12310a901a37489e550ebb133c/crates/evm/evm/src/executors/fuzz/mod.rs#L213-L222
> 
> Would be nice to extend this to invariant testing as well (would probably require the config to be a general one like `show_fuzz_logs` and apply to both fuzz and invariant) but can be handled in a separate PR.

hi @grandizzy 
After walking though the execution process. I don't think we could print the fuzz console logs right after  the executor call. here's why:
1. the console.log should also be controlled by `verbosity`, only `verbosity >= 2` is allowed to display logs. 
2. all logs should be printed under respective test results, which means that we have to keep the logs until it should be displayed.

related code:
https://github.com/foundry-rs/foundry/blob/8b694bbcabaedffc0337bf8dea9a135da5694ef9/crates/forge/bin/cmd/test/mod.rs#L455-L465

wdyt

**Comment by grandizzy** [comment]
> hi @grandizzy After walking though the execution process. I don't think we could print the fuzz console logs right after the executor call. here's why:
> 
>     1. the console.log should also be controlled by `verbosity`, only `verbosity >= 2` is allowed to display logs.

Yeah, I think the show fuzz logs config could supersede the verbosity, so don't think this is a biggie

> 
>     2. all logs should be printed under respective test results, which means that we have to keep the logs until it should be displayed.
> 

Good point, if we're not displaying under their own test then running multiple fuzz tests at the same time could become a mess. So let's stick with the impl of gather logs and displaying them at the end, let's keep for fuzz only for now and extend to invariants after. There's `FuzzTestData` that's already ref celled, so pls add that to it https://github.com/foundry-rs/foundry/blob/15652dc9ac56fa12310a901a37489e550ebb133c/crates/evm/evm/src/executors/fuzz/mod.rs#L25-L40

**Comment by Azleal** [comment]
> > hi @grandizzy After walking though the execution process. I don't think we could print the fuzz console logs right after the executor call. here's why:
> > ```
> > 1. the console.log should also be controlled by `verbosity`, only `verbosity >= 2` is allowed to display logs.
> > ```
> 
> Yeah, I think the show fuzz logs config could supersede the verbosity, so don't think this is a biggie
> 
> > ```
> > 2. all logs should be printed under respective test results, which means that we have to keep the logs until it should be displayed.
> > ```
> 
> Good point, if we're not displaying under their own test then running multiple fuzz tests at the same time could become a mess. So let's stick with the impl of gather logs and displaying them at the end, let's keep for fuzz only for now and extend to invariants after. There's `FuzzTestData` that's already ref celled, so pls add that to it (instead declaring the `logs: RefCell<Vec<Log>>`)
> 
> https://github.com/foundry-rs/foundry/blob/15652dc9ac56fa12310a901a37489e550ebb133c/crates/evm/evm/src/executors/fuzz/mod.rs#L25-L40

do you mean that I should define a new field like `logs` in `FuzzTestData` to store logs, or I could make use of `logs` in `counterexample` or `traces`?

**Comment by grandizzy** [comment]
> > > hi @grandizzy After walking though the execution process. I don't think we could print the fuzz console logs right after the executor call. here's why:
> > > ```
> > > 1. the console.log should also be controlled by `verbosity`, only `verbosity >= 2` is allowed to display logs.
> > > ```
> > 
> > 
> > Yeah, I think the show fuzz logs config could supersede the verbosity, so don't think this is a biggie
> > > ```
> > > 2. all logs should be printed under respective test results, which means that we have to keep the logs until it should be displayed.
> > > ```
> > 
> > 
> > Good point, if we're not displaying under their own test then running multiple fuzz tests at the same time could become a mess. So let's stick with the impl of gather logs and displaying them at the end, let's keep for fuzz only for now and extend to invariants after. There's `FuzzTestData` that's already ref celled, so pls add that to it (instead declaring the `logs: RefCell<Vec<Log>>`)
> > https://github.com/foundry-rs/foundry/blob/15652dc9ac56fa12310a901a37489e550ebb133c/crates/evm/evm/src/executors/fuzz/mod.rs#L25-L40
> 
> do you mean that I should define a new field like `logs` in `FuzzTestData` to store logs, or I could make use of `logs` in `counterexample` or `traces`?

Yeah, I mean new field in struct instead https://github.com/foundry-rs/foundry/pull/8387/files#diff-12784ea10f5753f595ab88ff54a8d09ecdd044d6a19676229f601a1bffe702e6R93

**Comment by Azleal** [comment]
> Thank you, nice progress!
> 
> I was thinking more along the lines of #3844 requirement, that is to print logs during the test rather than after the test is completed (good for long campaigns where one would like to get feedback during execution). So instead storing logs for all fuzz cases (which would also require more resources for executing these test) we can print logs right after the executor call (if `show-execution-logs` enabled), smth like adding
> 
> ```rust
>         if self.config.show_execution_logs {
>             for log in decode_console_logs(&call.logs) {
>                 println!("  {log}");
>             }
>         }
> ```
> 
> right after
> 
> https://github.com/foundry-rs/foundry/blob/15652dc9ac56fa12310a901a37489e550ebb133c/crates/evm/evm/src/executors/fuzz/mod.rs#L213-L222
> 
> Would be nice to extend this to invariant testing as well (would probably require the config to be a general one like `show_fuzz_logs` and apply to both fuzz and invariant) but can be handled in a separate PR.

@grandizzy Can we just use `show_logs` in `FuzzConfig`? Because I noticed that, there are `runs` in both `FuzzConfig` and `InvariantConfig`, and share the same meaning as runs for fuzz test or invariants test. wdyt

And do I need to add something like `fuzz_show_logs` here👇🏻 for users to use environment `FOUNDRY_FUZZ_SHOW_LOGS`?

https://github.com/foundry-rs/foundry/blob/0116be1bb3f73a6365290ba2894813815d02159d/crates/forge/bin/cmd/test/mod.rs#L105-L114

**Comment by grandizzy** [comment]
> @grandizzy Can we just use `show_logs` in `FuzzConfig`? Because I noticed that, there are `runs` in both `FuzzConfig` and `InvariantConfig`, and share the same meaning as runs for fuzz test or invariants test. wdyt

yes, that should work

> 
> And do I need to add something like `fuzz_show_logs` here👇🏻 for users to use environment `FOUNDRY_FUZZ_SHOW_LOGS`?
> 
nope, that's not needed

**Comment by Azleal** [comment]
> > @grandizzy Can we just use `show_logs` in `FuzzConfig`? Because I noticed that, there are `runs` in both `FuzzConfig` and `InvariantConfig`, and share the same meaning as runs for fuzz test or invariants test. wdyt
> 
> yes, that should work

well, I think this is my final commit. please review this to see if any change should be made. @grandizzy

File: crates/evm/evm/src/executors/fuzz/mod.rs [hunk 2]
```diff
@@ -113,6 +116,7 @@ impl FuzzedExecutor {
                         data.traces.push(call_traces);
                         data.breakpoints.replace(case.breakpoints);
                     }
+                    data.logs.extend(case.logs);
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/fuzz/mod.rs:119]: maybe extend only if `show_logs` is true? Then 
```Solidity
        // decide whether to use trace logs or only error logs
        let actual_logs = if show_logs { fuzz_result.logs } else { call.logs };
```

could go away and simply use `fuzz_result.logs`

File: crates/evm/evm/src/executors/fuzz/mod.rs [hunk 3]
```diff
@@ -133,6 +137,8 @@ impl FuzzedExecutor {
                     // to run at least one more case to find a minimal failure
                     // case.
                     let reason = rd.maybe_decode(&outcome.1.result, Some(status));
+                    let cloned_logs = outcome.1.logs.clone();
+                    execution_data.borrow_mut().logs.extend(cloned_logs);
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/fuzz/mod.rs:141]: let's do a one liner ```execution_data.borrow_mut().logs.extend(outcome.1.logs.clone());```

**Review by grandizzy — CHANGES_REQUESTED** [review]
looks good, think we can simplify a little bit, pls check

File: crates/evm/evm/src/executors/fuzz/mod.rs [hunk 4]
```diff
@@ -150,14 +156,17 @@ impl FuzzedExecutor {
...
+
         let mut result = FuzzTestResult {
             first_case: fuzz_result.first_case.unwrap_or_default(),
             gas_by_case: fuzz_result.gas_by_case,
             success: run_result.is_ok(),
             reason: None,
             counterexample: None,
-            decoded_logs: decode_console_logs(&call.logs),
-            logs: call.logs,
+            decoded_logs: decode_console_logs(&actual_logs),
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/fuzz/mod.rs:168]: I think we can get rid of `decode_logs` from `FuzzTestResult` as logs are decoded already when test result created

https://github.com/foundry-rs/foundry/blob/fdf165b65c7344415b684eb4ad8215ac7b713be2/crates/forge/src/result.rs#L493-L515

File: crates/evm/evm/src/executors/fuzz/mod.rs (hunk 2 above)
Comment (Azleal) [reply crates/evm/evm/src/executors/fuzz/mod.rs]: done. only when it's true that needs to be extended.

**Review by grandizzy — APPROVED** [review]
lgtm, thank you! clippy failure not related

**Comment by Azleal** [comment]
> lgtm, thank you! clippy failure not related

Thank you @grandizzy for your patience and guidance along this PR and you're very meticulous, 

> lgtm, thank you! clippy failure not related

Thank you, @grandizzy, for your patience and guidance throughout this PR. You're so meticulous. I'm new to Rust, and every time I thought there was nothing to optimize, your review showed me a better way. 😄👍🏻

Hopefully, this PR will not be reverted again.

**Review by mattsse — APPROVED** [review]
ty

**Pull request merged**